### PR TITLE
Enabled the IVF algorithm to work with Filters of K-NN Query.

### PR DESCRIPTION
**Pull request by navneet1v**
### Description
Enabled the IVF algorithm to work with Filters of K-NN Query.
 
### Issues Resolved
#903 

Current main branch build is broken. I will forward port that change once the issue is resolved.
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 1]
```diff
@@ -351,4 +353,18 @@ public static float normalizeScore(float score) {
...
+        if(StringUtils.isNotEmpty(spaceTypeString)) {
+            return SpaceType.getSpace(spaceTypeString);
+        }
+
+        final String modelId = fieldInfo.getAttribute(MODEL_ID);
+        if(StringUtils.isNotEmpty(modelId)) {
+            ModelMetadata modelMetadata = modelDao.getMetadata(modelId);
+            return modelMetadata.getSpaceType();
+        }
+        throw new IllegalArgumentException("Unable to find the Model Id from Field Info attributes for field " + fieldInfo.getName());
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/knn/index/query/KNNWeight.java:368]: please use ```String.format()```

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 2]
```diff
@@ -285,12 +286,13 @@ private Map<Integer, Float> doANNSearch(final LeafReaderContext context, final i
     }
 
     private Map<Integer, Float> doExactSearch(final LeafReaderContext leafReaderContext, final int[] filterIdsArray) {
+        log.error("Doing Exact search");
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/knn/index/query/KNNWeight.java:289]: is this right that logging is at ```error``` level?

Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:289]: This is interesting. I removed it from my local, but it is still coming. Let me double check.

**Review by martin-gaievski — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -15,6 +15,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 ## [Unreleased 2.x](https://github.com/opensearch-project/k-NN/compare/2.9...2.x)
 ### Features
 ### Enhancements
+1. Enabled the IVF algorithm to work with Filters of K-NN Query. [#1013](https://github.com/opensearch-project/k-NN/pull/1013)
```
Comment (jmazanec15) [inline CHANGELOG.md:18]: We typically use bullets not numbers?

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 4]
```diff
@@ -351,4 +353,20 @@ public static float normalizeScore(float score) {
...
+            return SpaceType.getSpace(spaceTypeString);
+        }
+
+        final String modelId = fieldInfo.getAttribute(MODEL_ID);
+        if (StringUtils.isNotEmpty(modelId)) {
+            ModelMetadata modelMetadata = modelDao.getMetadata(modelId);
+            return modelMetadata.getSpaceType();
+        }
+        throw new IllegalArgumentException(
+            String.format(Locale.ROOT, "Unable to find the Model Id from Field Info attributes for field %s", fieldInfo.getName())
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/query/KNNWeight.java:369]: Should this be something else?

Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:369]: Like what?

File: CHANGELOG.md (hunk 3 above)
Comment (navneet1v) [reply CHANGELOG.md:18]: I can change it to bullet. I just thoughts numbers might work. will fix this.

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java (hunk 4 above)
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:369]: Does this work: Unable to find the Space Type from Field Info attributes for field?

**Review by martin-gaievski — APPROVED** [review]

**Review by jmazanec15 — APPROVED** [review]

**Pull request merged**